devel/fortio/http.go

```diff
@@ -147,17 +161,486 @@ func NewClient(url string, numConnections int, compression bool) *Client {
+			}
+		} else {
+			// After digits, skipping ahead to find \r\n
+			if inp[off] == '\r' {
+				off++
+				if off >= end {
+					return off, -1
+				}
+				if inp[off] == '\n' {
+					// good case
+					return off + 1, res
+				}
+			}
+		}
+		off++
+	}
+}
+
+// return the result from the state.
+func (c *BasicClient) returnRes() (int, []byte, int) {
+	return c.code, c.buffer[:c.size], c.headerLen
+}
+
+// connect to destination.
+func (c *BasicClient) connect() *net.TCPConn {
+	socket, err := net.DialTCP("tcp", nil, &c.dest)
+	if err != nil {
+		Err("Unable to connect to %v : %v", c.dest, err)
+		return nil
+	}
+	// For now those errors are not critical/breaking
+	if err = socket.SetNoDelay(true); err != nil {
+		Warn("Unable to connect to set tcp no delay %v %v : %v", socket, c.dest, err)
+	}
+	if err = socket.SetWriteBuffer(len(c.req)); err != nil {
+		Warn("Unable to connect to set write buffer %d %v %v : %v", len(c.req), socket, c.dest, err)
+	}
+	if err = socket.SetReadBuffer(len(c.buffer)); err != nil {
+		Warn("Unable to connect to read buffer %d %v %v : %v", len(c.buffer), socket, c.dest, err)
+	}
+	return socket
+}
+
+// Fetch fetches the url content. Returns http code, data, offset of body.
+func (c *BasicClient) Fetch() (int, []byte, int) {
+	c.code = -1
+	c.size = 0
+	c.headerLen = 0
+	// Connect or reuse existing socket:
+	conn := c.socket
+	reuse := (conn != nil)
+	if !reuse {
+		conn = c.connect()
+		if conn == nil {
+			return c.returnRes()
+		}
+	} else {
+		Dbg("Reusing socket %v", *conn)
+	}
+	c.socket = nil // because of error returns
+	// Send the request:
+	n, err := conn.Write(c.req)
+	if err != nil {
+		if reuse {
+			// it's ok for the (idle) socket to die once, auto reconnect:
+			Info("Closing dead socket %v (%v)", *conn, err)
+			conn.Close() // nolint: errcheck
+			c.errorCount++
+			return c.Fetch() // recurse once
+		}
+		Err("Unable to write to %v %v : %v", conn, c.dest, err)
+		return c.returnRes()
+	}
+	if n != len(c.req) {
+		Err("Short write to %v %v : %d instead of %d", conn, c.dest, n, len(c.req))
+		return c.returnRes()
+	}
+	if !c.keepAlive {
+		if err = conn.CloseWrite(); err != nil {
+			Err("Unable to close write to %v %v : %v", conn, c.dest, err)
+			return c.returnRes()
+		}
+	}
+	// Read the response:
+	c.readResponse(conn)
+	// Return the result:
+	return c.returnRes()
+}
+
+// EscapeBytes returns printable string. Same as %q format without the
+// surrounding/extra "".
+func EscapeBytes(buf []byte) string {
+	e := fmt.Sprintf("%q", buf)
+	return e[1 : len(e)-1]
+}
+
+// DebugSummary returns a string with the size and escaped first max/2 and
+// last max/2 bytes of a buffer (or the whole escaped buffer if small enough).
+func DebugSummary(buf []byte, max int) string {
+	l := len(buf)
+	if l <= max+3 { //no point in shortening to add ... if we could return those 3
+		return EscapeBytes(buf)
+	}
+	max /= 2
+	return fmt.Sprintf("%d: %s...%s", l, EscapeBytes(buf[:max]), EscapeBytes(buf[l-max:]))
```

Comment:
it's mostly cosmetic/to save 4 characters on the line
Len: Beginning...End
is nicer to me than
Len: "Beginning"..."End"